GraphicsToolsProject/Plugins/GraphicsTools/Source/GraphicsToolsEditor/Private/GTMeshOutlineComponentDetails.cpp

```diff
@@ -0,0 +1,354 @@
+																		 "modified to work with mesh outlines. Does not modify instance."))
+								  .OnClicked(this, &FGTMeshOutlineComponentDetails::ClickedOnConvertToStaticMesh)
+								  .IsEnabled(this, &FGTMeshOutlineComponentDetails::ConvertToStaticMeshEnabled)
+								  .Content()[SNew(STextBlock).Text(ConvertToStaticMeshText)]];
+}
+
+UGTMeshOutlineComponent* FGTMeshOutlineComponentDetails::GetFirstSelectedMeshOutlineComponent() const
+{
+	UGTMeshOutlineComponent* OutlineMeshComponent = nullptr;
+
+	for (const TWeakObjectPtr<UObject>& Object : SelectedObjectsList)
+	{
+		UGTMeshOutlineComponent* TestOutlineMeshComponent = Cast<UGTMeshOutlineComponent>(Object.Get());
+
+		if (TestOutlineMeshComponent != nullptr && !TestOutlineMeshComponent->IsTemplate() &&
+			TestOutlineMeshComponent->GetStaticMesh() != nullptr)
+		{
+			OutlineMeshComponent = TestOutlineMeshComponent;
+			break;
+		}
+	}
+
+	return OutlineMeshComponent;
+}
+
+bool FGTMeshOutlineComponentDetails::ConvertToStaticMeshEnabled() const
+{
+	return GetFirstSelectedMeshOutlineComponent() != nullptr;
+}
+
+typedef TPair<int32, FVector> IndexVertex;
+
+void SmoothNormals(const TArray<FVector>& Vertices, TArray<FVector>& Normals, TArray<FProcMeshTangent>& Tangents)
+{
+	TMap<FVector, TArray<IndexVertex>> GroupedVerticies;
+	GroupedVerticies.Reserve(Vertices.Num()); // Reserve for the worse case, each vertex is at a unique location.
+
+	// Group all vertices that share the same location in space.
+	{
+		const int32 Count = Vertices.Num();
+		for (int32 Index = 0; Index < Count; ++Index)
+		{
+			TArray<IndexVertex>& SmoothingGroup = GroupedVerticies.FindOrAdd(Vertices[Index]);
+			SmoothingGroup.Add(IndexVertex(Index, Vertices[Index]));
+		}
+	}
+
+	// If we didn't hit the degenerate case of each vertex is its own group (no vertices shared a location), average the normals/tangents of
+	// each group.
+	if (GroupedVerticies.Num() != Vertices.Num())
+	{
+		TArray<FVector> SmoothNormals(Normals);
+		TArray<FProcMeshTangent> SmoothTangents(Tangents);
+
+		for (auto& Pair : GroupedVerticies)
+		{
+			TArray<IndexVertex>& SmoothingGroup = Pair.Value;
+
+			// No need to smooth a group of one.
+			if (SmoothingGroup.Num() != 1)
+			{
+				// Average the normals and tangents.
+				FVector SmoothedNormal = FVector::ZeroVector;
+				FVector SmoothedTangent = FVector::ZeroVector;
+
+				for (auto& Vertex : SmoothingGroup)
+				{
+					SmoothedNormal += Normals[Vertex.Key];
+					SmoothedTangent += Tangents[Vertex.Key].TangentX;
+				}
+
+				SmoothedNormal.Normalize();
+				SmoothedTangent.Normalize();
+
+				for (auto& Vertex : SmoothingGroup)
+				{
+					SmoothNormals[Vertex.Key] = SmoothedNormal;
+					SmoothTangents[Vertex.Key] = FProcMeshTangent(SmoothedTangent, false);
+				}
+			}
+		}
+
+		Normals = SmoothNormals;
+		Tangents = SmoothTangents;
+	}
+}
+
+bool BuildOutlineMesh(UProceduralMeshComponent* ProceduralMesh, UGTMeshOutlineComponent* OutlineMesh)
```

Comment:
>BuildOutlineMesh [](http://example.com/codeflow?start=5&length=16)

Out of curiosity, why can't we just use the original mesh and invert the normals in the vertex shader?

Reply:
Great question! Yes you can do that, and that was the initial route I went down. But... Unreal doesn't have an interface for enabling front-face culling. The options are back-face (the default) or no culling (two sided). So to emulate front-face culling you have to make your material a cutout material (alpha tested) and do pixel clipping based on the triangle's facing. This incurs some performance overhead since your material is no longer opaque and the renderer can no longer perform early depth testing (before the fragment stage). :( So I decided if I was going to smooth the normals for some meshes why not enforce that users should do the triangle flipping at edit time rather than run time.